pkg/traces/promsdprocessor/consumer/consumer.go

```diff
@@ -0,0 +1,256 @@
+package consumer
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/component/discovery"
+	"github.com/prometheus/common/model"
+	"go.opentelemetry.io/collector/client"
+	otelcomponent "go.opentelemetry.io/collector/component"
+	otelconsumer "go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+	semconv "go.opentelemetry.io/collector/semconv/v1.5.0"
+)
+
+const (
+	// OperationTypeInsert inserts a new k/v if it isn't already present
+	OperationTypeInsert = "insert"
+	// OperationTypeUpdate only modifies an existing k/v
+	OperationTypeUpdate = "update"
+	// OperationTypeUpsert does both of above
+	OperationTypeUpsert = "upsert"
+
+	//TODO: It'd be cleaner to get these from the otel semver package?
+	//      Not all are in semver though. E.g. "k8s.pod.ip" is internal inside the k8sattributesprocessor.
+	PodAssociationIPLabel       = "ip"
+	PodAssociationOTelIPLabel   = "net.host.ip"
+	PodAssociationk8sIPLabel    = "k8s.pod.ip"
+	PodAssociationHostnameLabel = "hostname"
+	PodAssociationConnectionIP  = "connection"
+)
+
+func ValidateOperationType(operationType string) error {
+	switch operationType {
+	case
+		OperationTypeUpsert,
+		OperationTypeInsert,
+		OperationTypeUpdate:
+		// Valid configuration, do nothing.
+	default:
+		return fmt.Errorf("unknown operation type %s", operationType)
+	}
+	return nil
+}
+
+func ValidatePodAssociations(podAssociations []string) error {
+	for _, podAssociation := range podAssociations {
+		switch podAssociation {
+		case
+			PodAssociationIPLabel,
+			PodAssociationOTelIPLabel,
+			PodAssociationk8sIPLabel,
+			PodAssociationHostnameLabel,
+			PodAssociationConnectionIP:
+			// Valid configuration, do nothing.
+		default:
+			return fmt.Errorf("unknown pod association %s", podAssociation)
+		}
+	}
+	return nil
+}
+
+// TODO: Put a private member so that this can't be created without calling NewConsumer?
+type Consumer struct {
+	optsMut sync.RWMutex
+	opts    Options
+
+	logger log.Logger
+}
+
+// Options configure a Consumer
+type Options struct {
+	HostLabels      map[string]discovery.Target
+	OperationType   string
+	PodAssociations []string
+	NextConsumer    otelconsumer.Traces
+}
+
+var _ otelconsumer.Traces = (*Consumer)(nil)
+
+func NewConsumer(opts Options, logger log.Logger) (*Consumer, error) {
+	c := &Consumer{
+		logger: logger,
+	}
+
+	err := c.UpdateOptions(opts)
```

Comment:
@mattdurham I updated `NewConsumer` to call `UpdateOptions` instead of setting the options directly. This ensures we do the same checks on new options in both `NewConsumer` and in `UpdateOptions`.